compiler/aarch64/codegen/ARM64BinaryEncoding.cpp

```diff
@@ -0,0 +1,235 @@
+/*******************************************************************************
+ * Copyright (c) 2018, 2018 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "codegen/ARM64ConditionCode.hpp"
+#include "codegen/ARM64Instruction.hpp"
+#include "codegen/CodeGenerator.hpp"
+
+uint8_t *OMR::ARM64::Instruction::generateBinaryEncoding()
+   {
+   uint8_t *instructionStart = self()->cg()->getBinaryBufferCursor();
+   uint8_t *cursor = instructionStart;
+   cursor = self()->getOpCode().copyBinaryToBuffer(instructionStart);
+   cursor += ARM64_INSTRUCTION_LENGTH;
+   self()->setBinaryLength(ARM64_INSTRUCTION_LENGTH);
+   self()->setBinaryEncoding(instructionStart);
+   return cursor;
+   }
+
+int32_t OMR::ARM64::Instruction::estimateBinaryLength(int32_t currentEstimate)
+   {
+   self()->setEstimatedBinaryLength(ARM64_INSTRUCTION_LENGTH);
+   return currentEstimate + self()->getEstimatedBinaryLength();
```

Comment:
This part of the instruction hierarchy isn't extensible, so `self()` usage in this function isn't strictly correct.  Please remove.